oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2AuthorizationCodeAuthenticationProvider.java

```diff
@@ -117,24 +124,36 @@ public Authentication authenticate(Authentication authentication) throws Authent
 		}
 
 		Set<String> authorizedScopes = authorization.getAttribute(OAuth2AuthorizationAttributeNames.AUTHORIZED_SCOPES);
-		Jwt jwt = OAuth2TokenIssuerUtil
-			.issueJwtAccessToken(this.jwtEncoder, authorization.getPrincipalName(), registeredClient.getClientId(), authorizedScopes);
-		OAuth2AccessToken accessToken = new OAuth2AccessToken(OAuth2AccessToken.TokenType.BEARER,
-				jwt.getTokenValue(), jwt.getIssuedAt(), jwt.getExpiresAt(), authorizedScopes);
+		OAuth2TokenRequest accessTokenRequest = OAuth2TokenRequest.builder()
+				.scopes(authorizedScopes)
+				.registeredClient(registeredClient)
+				.resourceOwner(authorization.getPrincipalName())
+				.timeToLive(Duration.ofMinutes(60)) // TODO: make this configurable
+				.build();
 
+		OAuth2TokenResult<OAuth2AccessToken> tokenResult = this.accessTokenIssuer.issue(accessTokenRequest);
+		OAuth2AccessToken accessToken = tokenResult.getToken();
 		OAuth2Tokens.Builder tokensBuilder = OAuth2Tokens.from(authorization.getTokens())
-				.accessToken(accessToken);
+				.accessToken(accessToken, tokenResult.getMetadata());
 
 		OAuth2RefreshToken refreshToken = null;
 		if (registeredClient.getTokenSettings().enableRefreshTokens()) {
-			refreshToken = OAuth2TokenIssuerUtil.issueRefreshToken(registeredClient.getTokenSettings().refreshTokenTimeToLive());
-			tokensBuilder.refreshToken(refreshToken);
+			OAuth2TokenRequest refreshTokenRequest = OAuth2TokenRequest.builder()
+					.scopes(authorizedScopes)
+					.registeredClient(registeredClient)
+					.resourceOwner(authorization.getPrincipalName())
+					.timeToLive(registeredClient.getTokenSettings().refreshTokenTimeToLive())
+					.build();
+
+			OAuth2TokenResult<OAuth2RefreshToken> refreshTokenResult = refreshTokenIssuer.issue(refreshTokenRequest);
+			refreshToken = refreshTokenResult.getToken();
+			tokensBuilder.refreshToken(refreshToken, refreshTokenResult.getMetadata());
 		}
 
 		OAuth2Tokens tokens = tokensBuilder.build();
 		authorization = OAuth2Authorization.from(authorization)
 				.tokens(tokens)
-				.attribute(OAuth2AuthorizationAttributeNames.ACCESS_TOKEN_ATTRIBUTES, jwt)
+				.attribute(OAuth2AuthorizationAttributeNames.ACCESS_TOKEN_METADATA, tokenResult.getMetadata())
```

Comment:
The metadata is already associated in `OAuth2Tokens` so no need to store here.

Reply:
Good point. Do you think anything should go to `OAuth2Authorization` attributes at all here? Token metadata now holds originally generated `Jwt` object, so there is no need to store it here.

Reply:
Going to remove this attribute completely since the the original token is now in `OAuth2TokenMetadata`.